copts.bzl

```diff
@@ -14,26 +14,145 @@
 #  limitations under the License.
 #
 
-GCC_FLAGS = [
-        "-pthread",
-        "-std=c++11",
-#        "-Werror",
-#        "-Wsign-compare",
-#        "-Wno-unused-parameter",
-#        "-Wno-unused-variable",
-#        "-Woverloaded-virtual",
-#        "-Wnon-virtual-dtor",
-#        "-Wno-missing-field-initializers",
-]
-
-GCC_TEST_FLAGS = [
-        "-pthread",
-        "-std=c++11",
-#        "-Werror",
-#        "-Wsign-compare",
-#        "-Wno-unused-parameter",
-#        "-Wno-unused-variable",
-#        "-Woverloaded-virtual",
-#        "-Wnon-virtual-dtor",
-#        "-Wno-missing-field-initializers",
+#
+# Copyright 2017 The Abseil Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+BASE_FLAGS = [
+    "-DOS_LINUX",
+    "-DSNAPPY",
+    "-DHAVE_SSE42",
+    "-fno-omit-frame-pointer",
+    "-momit-leaf-frame-pointer",
+    "-msse4.2",
+    "-pthread",
+]
+
+CXX_FLAGS = [
+    "-std=c++11",
+]
+
+CURVE_GCC_FLAGS = [
+    "-Wall",
+    "-Wextra",
+    "-Wconversion-null",
+    "-Wformat-security",
+    "-Woverlength-strings",
+    "-Wpointer-arith",
+    "-Wundef",
+    "-Wunused-local-typedefs",
+    "-Wunused-result",
+    "-Wvarargs",
+    "-Wvla",
+    "-Wwrite-strings",
+    "-Werror",
+    "-DNOMINMAX",
+]
+
+CURVE_GCC_TEST_FLAGS = [
+    "-Wno-conversion-null",
+    "-Wno-deprecated-declarations",
+    "-Wno-missing-declarations",
+    "-Wno-sign-compare",
+    "-Wno-unused-function",
+    "-Wno-unused-parameter",
+    "-Wno-unused-result",
 ]
+
+# FIXME: temporary disabled because triggered in many places
+CURVE_GCC_DISABLED_FLGAS = [
+    "-Wno-error=sign-compare",
+    "-Wno-error=reorder",
+    "-Wno-error=unused-parameter",
+    "-Wno-error=unused-variable",
+    "-Wno-error=deprecated-declarations",
+    "-Wno-error=narrowing",
+]
+
+# FIXME: temporary disabled because triggered in many places
+CURVE_GCC_TEST_DISABLED_FLAGS = [
+    "-Wno-error=unused-but-set-variable",
+    "-Wno-error=vla",
+    "-Wno-error=uninitialized",
+    "-Wno-error=maybe-uninitialized",
+    "-Wno-error=format",
+    "-Wno-error=write-strings",
+    "-Wno-error=missing-field-initializers",
+]
+
+CURVE_LLVM_FLAGS = [
```

Comment:
What is this for？

Reply:
clang/llvm is another well-known c/c++ compiler, sometimes, it can give more informative error messages and generate a more efficient binary.
currently, this flags is not fully tested, because it failed when compiling brpc.